exception on pre-fulu coming from MinCustodyPeriodSlotCalculator

```
2025-11-25 12:10:00.033+10:00 | beaconchain-async-2 | ERROR | DataColumnSidecarCustodyImpl | Unexpected error while advancing first custody incomplete slot for checkpoint Checkpoint{epoch=2, root=0xd4515640459e85c0bbce43e646456a07870d2f4d6ecfcefdbc6d012ab024805a}
java.util.concurrent.CompletionException: java.util.NoSuchElementException: No value present
	at java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:332) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:347) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:636) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2194) ~[?:?]
	at tech.pegasys.teku.infrastructure.async.stream.AsyncIteratorCollector.onError(AsyncIteratorCollector.java:44) ~[teku-infrastructure-async-develop.jar:25.11.0+53-gc325ae9c13]
	at tech.pegasys.teku.infrastructure.async.stream.AbstractDelegatingStreamHandler.onError(AbstractDelegatingStreamHandler.java:31) ~[teku-infrastructure-async-develop.jar:25.11.0+53-gc325ae9c13]
	at tech.pegasys.teku.infrastructure.async.stream.AbstractDelegatingStreamHandler.onError(AbstractDelegatingStreamHandler.java:31) ~[teku-infrastructure-async-develop.jar:25.11.0+53-gc325ae9c13]
	at tech.pegasys.teku.infrastructure.async.stream.MapStreamHandler.onNext(MapStreamHandler.java:33) ~[teku-infrastructure-async-develop.jar:25.11.0+53-gc325ae9c13]
	at tech.pegasys.teku.infrastructure.async.stream.FutureAsyncIteratorImpl.lambda$iterate$1(FutureAsyncIteratorImpl.java:30) ~[teku-infrastructure-async-develop.jar:25.11.0+53-gc325ae9c13]
	at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$finish$39(SafeFuture.java:505) ~[teku-infrastructure-async-develop.jar:25.11.0+53-gc325ae9c13]
	at java.base/java.util.concurrent.CompletableFuture.uniHandle(CompletableFuture.java:934) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture$UniHandle.tryFire(CompletableFuture.java:911) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2179) ~[?:?]
	at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateResult$2(SafeFuture.java:150) ~[teku-infrastructure-async-develop.jar:25.11.0+53-gc325ae9c13]
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:863) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:841) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:510) ~[?:?]
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2179) ~[?:?]
	at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateToAsync$29(SafeFuture.java:461) ~[teku-infrastructure-async-develop.jar:25.11.0+53-gc325ae9c13]
	at tech.pegasys.teku.infrastructure.async.SafeFuture.of(SafeFuture.java:82) ~[teku-infrastructure-async-develop.jar:25.11.0+53-gc325ae9c13]
	at tech.pegasys.teku.infrastructure.async.AsyncRunner.lambda$runAsync$2(AsyncRunner.java:47) ~[teku-infrastructure-async-develop.jar:25.11.0+53-gc325ae9c13]
	at tech.pegasys.teku.infrastructure.async.SafeFuture.of(SafeFuture.java:74) ~[teku-infrastructure-async-develop.jar:25.11.0+53-gc325ae9c13]
	at tech.pegasys.teku.infrastructure.async.ScheduledExecutorAsyncRunner.lambda$createRunnableForAction$1(ScheduledExecutorAsyncRunner.java:124) ~[teku-infrastructure-async-develop.jar:25.11.0+53-gc325ae9c13]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
	at java.base/java.lang.Thread.run(Thread.java:1583) [?:?]
Caused by: java.util.NoSuchElementException: No value present
	at java.base/java.util.Optional.orElseThrow(Optional.java:377) ~[?:?]
	at tech.pegasys.teku.statetransition.datacolumns.MinCustodyPeriodSlotCalculator.lambda$createFromSpec$0(MinCustodyPeriodSlotCalculator.java:31) ~[teku-ethereum-statetransition-develop.jar:25.11.0+53-gc325ae9c13]
	at tech.pegasys.teku.statetransition.datacolumns.DataColumnSidecarCustodyImpl.lambda$retrievePotentiallyIncompleteSlotCustodies$5(DataColumnSidecarCustodyImpl.java:238) ~[teku-ethereum-statetransition-develop.jar:25.11.0+53-gc325ae9c13]
	at java.base/java.util.Optional.orElseGet(Optional.java:364) ~[?:?]
	at tech.pegasys.teku.statetransition.datacolumns.DataColumnSidecarCustodyImpl.lambda$retrievePotentiallyIncompleteSlotCustodies$6(DataColumnSidecarCustodyImpl.java:236) ~[teku-ethereum-statetransition-develop.jar:25.11.0+53-gc325ae9c13]
	at tech.pegasys.teku.infrastructure.async.stream.MapStreamHandler.onNext(MapStreamHandler.java:31) ~[teku-infrastructure-async-develop.jar:25.11.0+53-gc325ae9c13]
	... 19 more
```

--- a/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/DasCustodySync.java
+++ b/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/DasCustodySync.java
@@ -151,8 +151,9 @@ public class DasCustodySync implements SlotEventsChannel {
                 if (missingColumn
                     .slot()
                     .isGreaterThan(
-                        minCustodyPeriodSlotCalculator.getMinCustodyPeriodSlot(
-                            currentSlot.get()))) {
+                        minCustodyPeriodSlotCalculator
+                            .getMinCustodyPeriodSlot(currentSlot.get())
+                            .orElseThrow())) {
                   addPendingRequest(missingColumn);
                 } else {
                   LOG.debug(

--- a/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/DataColumnSidecarCustodyImpl.java
+++ b/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/DataColumnSidecarCustodyImpl.java
@@ -167,15 +167,17 @@ public class DataColumnSidecarCustodyImpl
 
     LOG.debug(
         "Custody group count increased from {} to {}", oldCustodyGroupCount, newCustodyGroupCount);
-    final UInt64 minCustodyPeriodSlot =
-        minCustodyPeriodSlotCalculator.getMinCustodyPeriodSlot(slot);
-    db.setFirstCustodyIncompleteSlot(minCustodyPeriodSlot)
-        .finish(
-            error ->
-                LOG.error(
-                    "Unexpected error while updating first custody incomplete slot with a new value: {}.",
-                    minCustodyPeriodSlot,
-                    error));
+    minCustodyPeriodSlotCalculator
+        .getMinCustodyPeriodSlot(slot)
+        .ifPresent(
+            minCustodyPeriodSlot ->
+                db.setFirstCustodyIncompleteSlot(minCustodyPeriodSlot)
+                    .finish(
+                        error ->
+                            LOG.error(
+                                "Unexpected error while updating first custody incomplete slot with a new value: {}.",
+                                minCustodyPeriodSlot,
+                                error)));
   }
 
   @Override
@@ -232,13 +234,16 @@ public class DataColumnSidecarCustodyImpl
     return AsyncStream.create(db.getFirstCustodyIncompleteSlot())
         .flatMap(
             maybeFirstIncompleteSlot -> {
-              final UInt64 firstIncompleteSlot =
-                  maybeFirstIncompleteSlot.orElseGet(
+              final Optional<UInt64> firstIncompleteSlot =
+                  maybeFirstIncompleteSlot.or(
                       () ->
                           minCustodyPeriodSlotCalculator.getMinCustodyPeriodSlot(
                               currentSlot.get()));
+              if (firstIncompleteSlot.isEmpty()) {
+                return AsyncStream.empty();
+              }
               final Stream<UInt64> slotStream =
-                  UInt64.rangeClosed(firstIncompleteSlot, toSlotIncluded);
+                  UInt64.rangeClosed(firstIncompleteSlot.get(), toSlotIncluded);
               return AsyncStream.createUnsafe(slotStream.iterator())
                   .mapAsync(this::retrieveSlotCustody);
             });
@@ -247,7 +252,12 @@ public class DataColumnSidecarCustodyImpl
   @VisibleForTesting
   SafeFuture<SlotCustody> retrieveSlotCustody(final UInt64 slot) {
     if (slot.isLessThan(
-        minCustodyPeriodSlotCalculator.getMinCustodyPeriodSlot(currentSlot.get()))) {
+        minCustodyPeriodSlotCalculator
+            .getMinCustodyPeriodSlot(currentSlot.get())
+            .orElseThrow(
+                () ->
+                    new IllegalStateException(
+                        "Cannot retrieve slot custody outside custody period (" + slot + ")")))) {
       LOG.trace(
           "Skipping custody for slot {}, because currentSlot {} is beyond minCustodyPeriod",
           slot,

--- a/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/MinCustodyPeriodSlotCalculator.java
+++ b/ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/MinCustodyPeriodSlotCalculator.java
@@ -13,9 +13,11 @@
 
 package tech.pegasys.teku.statetransition.datacolumns;
 
+import java.util.Optional;
 import tech.pegasys.teku.infrastructure.unsigned.UInt64;
 import tech.pegasys.teku.spec.Spec;
 import tech.pegasys.teku.spec.SpecMilestone;
+import tech.pegasys.teku.spec.config.SpecConfigFulu;
 
 public interface MinCustodyPeriodSlotCalculator {
 
@@ -25,17 +27,15 @@ public interface MinCustodyPeriodSlotCalculator {
 
     return currentSlot -> {
       final UInt64 currentEpoch = spec.computeEpochAtSlot(currentSlot);
-      final int custodyPeriodEpochs =
-          spec.getSpecConfig(currentEpoch)
-              .toVersionFulu()
-              .orElseThrow()
-              .getMinEpochsForDataColumnSidecarsRequests();
-
-      final UInt64 minCustodyEpoch =
-          currentEpoch.minusMinZero(custodyPeriodEpochs).max(fuluActivationEpoch);
-      return spec.computeStartSlotAtEpoch(minCustodyEpoch);
+      return spec.getSpecConfig(currentEpoch)
+          .toVersionFulu()
+          .map(SpecConfigFulu::getMinEpochsForDataColumnSidecarsRequests)
+          .map(
+              custodyPeriodEpochs ->
+                  currentEpoch.minusMinZero(custodyPeriodEpochs).max(fuluActivationEpoch))
+          .map(spec::computeStartSlotAtEpoch);
     };
   }
 
-  UInt64 getMinCustodyPeriodSlot(UInt64 currentSlot);
+  Optional<UInt64> getMinCustodyPeriodSlot(UInt64 currentSlot);
 }

--- a/ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/datacolumns/DasCustodySyncTest.java
+++ b/ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/datacolumns/DasCustodySyncTest.java
@@ -92,7 +92,8 @@ public class DasCustodySyncTest {
 
   @BeforeEach
   public void setup() {
-    when(minCustodyPeriodSlotCalculator.getMinCustodyPeriodSlot(any())).thenReturn(UInt64.ZERO);
+    when(minCustodyPeriodSlotCalculator.getMinCustodyPeriodSlot(any()))
+        .thenReturn(Optional.of(UInt64.ZERO));
   }
 
   @Test
@@ -280,7 +281,7 @@ public class DasCustodySyncTest {
     final DataColumnSidecarRetriever retriever = mock(DataColumnSidecarRetriever.class);
     final DataColumnSidecarCustody custody = mock(DataColumnSidecarCustody.class);
     when(minCustodyPeriodSlotCalculator.getMinCustodyPeriodSlot(any()))
-        .thenReturn(UInt64.valueOf(1024));
+        .thenReturn(Optional.of(UInt64.valueOf(1024)));
     when(custody.retrieveMissingColumns()).thenReturn(testData(columns));
     final DasCustodySync custodySync =
         new DasCustodySync(
@@ -371,7 +372,8 @@ public class DasCustodySyncTest {
 
   private void assertAllCustodyColumnsPresent() {
     assertCustodyColumnsPresent(
-        custodyStand.getMinCustodySlot().intValue(), custodyStand.getCurrentSlot().intValue());
+        custodyStand.getMinCustodySlot().orElseThrow().intValue(),
+        custodyStand.getCurrentSlot().intValue());
   }
 
   private void assertCustodyColumnsPresent(final int fromSlot, final int tillSlot) {

--- a/ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/datacolumns/DataColumnSidecarCustodyImplTest.java
+++ b/ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/datacolumns/DataColumnSidecarCustodyImplTest.java
@@ -14,6 +14,7 @@
 package tech.pegasys.teku.statetransition.datacolumns;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.verify;
@@ -55,8 +56,8 @@ import tech.pegasys.teku.statetransition.datacolumns.db.DataColumnSidecarDbAcces
 
 @SuppressWarnings({"FutureReturnValueIgnored"})
 public class DataColumnSidecarCustodyImplTest {
-
-  final Spec spec = TestSpecFactory.createMinimalFulu();
+  static final int FULU_ACTIVATION_SLOT = 8;
+  final Spec spec = TestSpecFactory.createMinimalWithFuluForkEpoch(UInt64.ONE);
   final DataColumnSidecarDB db = new DataColumnSidecarDBStub();
   final DataColumnSidecarDbAccessor dbAccessor =
       DataColumnSidecarDbAccessor.builder(db).spec(spec).build();
@@ -75,11 +76,13 @@ public class DataColumnSidecarCustodyImplTest {
 
   public static Stream<Arguments> minCustodyScenarios() {
     return Stream.of(
-        Arguments.of(0, Optional.of(ZERO)),
-        Arguments.of(32768, Optional.of(ZERO)),
-        Arguments.of(32765, Optional.of(ZERO)),
-        Arguments.of(32776, Optional.of(UInt64.valueOf(8))),
-        Arguments.of(32784, Optional.of(UInt64.valueOf(16))));
+        Arguments.of(0, Optional.empty()), // pre fulu
+        Arguments.of(32768 + FULU_ACTIVATION_SLOT, Optional.of(ZERO.plus(FULU_ACTIVATION_SLOT))),
+        Arguments.of(32765 + FULU_ACTIVATION_SLOT, Optional.of(ZERO.plus(FULU_ACTIVATION_SLOT))),
+        Arguments.of(
+            32776 + FULU_ACTIVATION_SLOT, Optional.of(UInt64.valueOf(8 + FULU_ACTIVATION_SLOT))),
+        Arguments.of(
+            32784 + FULU_ACTIVATION_SLOT, Optional.of(UInt64.valueOf(16 + FULU_ACTIVATION_SLOT))));
   }
 
   @BeforeEach
@@ -154,6 +157,7 @@ public class DataColumnSidecarCustodyImplTest {
 
   @Test
   public void onNewFinalizedCheckpoint_noFinalizedColumnsAvailable() {
+    custody.onSlot(UInt64.valueOf(FULU_ACTIVATION_SLOT));
     final SafeFuture<Void> future = custody.advanceFirstIncompleteSlot(UInt64.valueOf(2));
     assertThat(future).isCompleted();
     // if epoch 2 is final, then slots 0-23 are 'final' and slot 24 is the first non final
@@ -205,35 +209,47 @@ public class DataColumnSidecarCustodyImplTest {
   }
 
   @Test
+  public void retrieveSlotCustody_shouldThrowPreFulu() {
+    assertThatThrownBy(() -> custody.retrieveSlotCustody(ONE))
+        .isInstanceOf(IllegalStateException.class)
+        .hasMessage("Cannot retrieve slot custody outside custody period (1)");
+  }
+
+  @Test
   public void retrieveSlotCustody_insufficientCustody()
       throws ExecutionException, InterruptedException, TimeoutException {
     final CanonicalBlockResolver resolver = mock(CanonicalBlockResolver.class);
     final DataColumnSidecarDbAccessor sidecarDb = mock(DataColumnSidecarDbAccessor.class);
-    final BeaconBlock beaconBlock = dataStructureUtil.randomBeaconBlock(ONE);
+
+    final UInt64 fuluSlot = UInt64.valueOf(FULU_ACTIVATION_SLOT);
+    final BeaconBlock beaconBlock = dataStructureUtil.randomBeaconBlock(fuluSlot);
     final DataColumnIdentifier dataColumnIdentifier =
         new DataColumnIdentifier(beaconBlock.getRoot(), UInt64.ZERO);
-    when(resolver.getBlockAtSlot(ONE))
+    when(resolver.getBlockAtSlot(fuluSlot))
         .thenReturn(SafeFuture.completedFuture(Optional.of(beaconBlock)));
-    when(sidecarDb.getColumnIdentifiers(ONE))
+    when(sidecarDb.getColumnIdentifiers(fuluSlot))
         .thenReturn(
             SafeFuture.completedFuture(
-                List.of(new DataColumnSlotAndIdentifier(ONE, dataColumnIdentifier))));
+                List.of(new DataColumnSlotAndIdentifier(fuluSlot, dataColumnIdentifier))));
     when(custodyGroupCountManager.getCustodyColumnIndices()).thenReturn(List.of(ZERO, ONE));
     custody =
         new DataColumnSidecarCustodyImpl(
             spec, resolver, sidecarDb, minCustodyPeriodSlotCalculator, custodyGroupCountManager);
 
-    SafeFuture<DataColumnSidecarCustodyImpl.SlotCustody> future = custody.retrieveSlotCustody(ONE);
-    verify(sidecarDb).getColumnIdentifiers(ONE);
+    custody.onSlot(UInt64.valueOf(32)); // transition to fulu
+
+    SafeFuture<DataColumnSidecarCustodyImpl.SlotCustody> future =
+        custody.retrieveSlotCustody(fuluSlot);
+    verify(sidecarDb).getColumnIdentifiers(fuluSlot);
     verifyNoMoreInteractions(sidecarDb);
     final DataColumnSidecarCustodyImpl.SlotCustody slotCustody =
         future.get(100, TimeUnit.MILLISECONDS);
 
-    assertThat(slotCustody.slot()).isEqualTo(ONE);
+    assertThat(slotCustody.slot()).isEqualTo(fuluSlot);
     assertThat(slotCustody.requiredColumnIndices()).isEqualTo(List.of(ZERO, ONE));
     assertThat(slotCustody.getIncompleteColumns())
         .containsExactly(
-            new DataColumnSlotAndIdentifier(ONE, beaconBlock.getRoot(), UInt64.valueOf(1)));
+            new DataColumnSlotAndIdentifier(fuluSlot, beaconBlock.getRoot(), UInt64.valueOf(1)));
   }
 
   @Test

--- a/ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/datacolumns/DataColumnSidecarCustodyImplTest.java
+++ b/ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/datacolumns/DataColumnSidecarCustodyImplTest.java
@@ -145,14 +145,17 @@ public class DataColumnSidecarCustodyImplTest {
     custody.onSlot(UInt64.valueOf(1));
     assertThat(custody.getTotalCustodyGroupCount()).isEqualTo(groupCount);
     verifyNoInteractions(custodyGroupCountManager);
+    assertThat(dbAccessor.getFirstCustodyIncompleteSlot()).isCompletedWithValue(Optional.empty());
   }
 
   @Test
   public void onSlot_checksOnEpochBoundarySlot() {
     when(custodyGroupCountManager.getCustodyGroupCount()).thenReturn(groupCount + 3);
-    custody.onSlot(UInt64.valueOf(8));
+    custody.onSlot(UInt64.valueOf(FULU_ACTIVATION_SLOT));
     assertThat(custody.getTotalCustodyGroupCount()).isEqualTo(groupCount + 3);
     verify(custodyGroupCountManager).getCustodyGroupCount();
+    assertThat(dbAccessor.getFirstCustodyIncompleteSlot())
+        .isCompletedWithValue(Optional.of(UInt64.valueOf(8)));
   }
 
   @Test
@@ -163,6 +166,15 @@ public class DataColumnSidecarCustodyImplTest {
     // if epoch 2 is final, then slots 0-23 are 'final' and slot 24 is the first non final
     assertThat(dbAccessor.getFirstCustodyIncompleteSlot())
         .isCompletedWithValue(Optional.of(UInt64.valueOf(24)));
+  }
+
+  @Test
+  public void onNewFinalizedCheckpoint_doesNothingPreFulu() {
+    custody.onSlot(UInt64.valueOf(FULU_ACTIVATION_SLOT - 1));
+    final SafeFuture<Void> future = custody.advanceFirstIncompleteSlot(UInt64.valueOf(2));
+    assertThat(future).isCompleted();
+    // if epoch 2 is final, then slots 0-23 are 'final' and slot 24 is the first non final
+    assertThat(dbAccessor.getFirstCustodyIncompleteSlot()).isCompletedWithValue(Optional.empty());
   }
 
   @ParameterizedTest

--- a/ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/datacolumns/DataColumnSidecarCustodyImplTest.java
+++ b/ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/datacolumns/DataColumnSidecarCustodyImplTest.java
@@ -75,11 +75,11 @@ public class DataColumnSidecarCustodyImplTest {
 
   public static Stream<Arguments> minCustodyScenarios() {
     return Stream.of(
-        Arguments.of(0, 0),
-        Arguments.of(32768, 0),
-        Arguments.of(32765, 0),
-        Arguments.of(32776, 8),
-        Arguments.of(32784, 16));
+        Arguments.of(0, Optional.of(ZERO)),
+        Arguments.of(32768, Optional.of(ZERO)),
+        Arguments.of(32765, Optional.of(ZERO)),
+        Arguments.of(32776, Optional.of(UInt64.valueOf(8))),
+        Arguments.of(32784, Optional.of(UInt64.valueOf(16))));
   }
 
   @BeforeEach
@@ -163,9 +163,10 @@ public class DataColumnSidecarCustodyImplTest {
 
   @ParameterizedTest
   @MethodSource("minCustodyScenarios")
-  public void computesCustodyPeriodCorrectly(final int currentSlot, final int expectedCustodySlot) {
+  public void computesCustodyPeriodCorrectly(
+      final int currentSlot, final Optional<UInt64> expectedCustodySlot) {
     assertThat(minCustodyPeriodSlotCalculator.getMinCustodyPeriodSlot(UInt64.valueOf(currentSlot)))
-        .isEqualTo(UInt64.valueOf(expectedCustodySlot));
+        .isEqualTo(expectedCustodySlot);
   }
 
   @Test

--- a/ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/datacolumns/MinCustodyPeriodSlotCalculatorTest.java
+++ b/ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/datacolumns/MinCustodyPeriodSlotCalculatorTest.java
@@ -14,98 +14,79 @@
 package tech.pegasys.teku.statetransition.datacolumns;
 
 import static org.assertj.core.api.Assertions.assertThat;
-import static org.mockito.ArgumentMatchers.any;
-import static org.mockito.Mockito.mock;
-import static org.mockito.Mockito.when;
 
 import java.util.Optional;
-import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 import tech.pegasys.teku.infrastructure.unsigned.UInt64;
-import tech.pegasys.teku.spec.ForkSchedule;
 import tech.pegasys.teku.spec.Spec;
 import tech.pegasys.teku.spec.SpecMilestone;
-import tech.pegasys.teku.spec.config.SpecConfig;
-import tech.pegasys.teku.spec.config.SpecConfigFulu;
-import tech.pegasys.teku.spec.datastructures.state.Fork;
+import tech.pegasys.teku.spec.TestSpecFactory;
 
 class MinCustodyPeriodSlotCalculatorTest {
+  private final UInt64 fuluActivationEpoch = UInt64.valueOf(100);
 
-  private final Spec spec = mock(Spec.class);
-  private final ForkSchedule forkSchedule = mock(ForkSchedule.class);
-  private final Fork fuluFork = mock(Fork.class);
-  private final SpecConfig specConfig = mock(SpecConfig.class);
-  private final SpecConfigFulu specConfigFulu = mock(SpecConfigFulu.class);
-
-  private static final UInt64 SLOTS_PER_EPOCH = UInt64.valueOf(32);
-  private static final UInt64 FULU_ACTIVATION_EPOCH = UInt64.valueOf(100);
-  private static final int CUSTODY_PERIOD_EPOCHS = 10;
-
-  @BeforeEach
-  void setUp() {
-    when(spec.getForkSchedule()).thenReturn(forkSchedule);
-    when(forkSchedule.getFork(SpecMilestone.FULU)).thenReturn(fuluFork);
-    when(fuluFork.getEpoch()).thenReturn(FULU_ACTIVATION_EPOCH);
-
-    when(spec.getSpecConfig(any(UInt64.class))).thenReturn(specConfig);
-
-    when(specConfig.toVersionFulu()).thenReturn(Optional.of(specConfigFulu));
-    when(specConfigFulu.getMinEpochsForDataColumnSidecarsRequests())
-        .thenReturn(CUSTODY_PERIOD_EPOCHS);
-
-    when(spec.computeEpochAtSlot(any(UInt64.class)))
-        .thenAnswer(
-            invocation -> {
-              UInt64 slot = invocation.getArgument(0);
-              return slot.dividedBy(SLOTS_PER_EPOCH);
-            });
-
-    when(spec.computeStartSlotAtEpoch(any(UInt64.class)))
-        .thenAnswer(
-            invocation -> {
-              UInt64 epoch = invocation.getArgument(0);
-              return epoch.times(SLOTS_PER_EPOCH);
-            });
-  }
+  private final Spec spec = TestSpecFactory.createMinimalWithFuluForkEpoch(fuluActivationEpoch);
+  private final int slotsPerEpoch = spec.getGenesisSpec().getSlotsPerEpoch();
+  private final int minEpochsForDataColumnSidecarsRequests =
+      spec.forMilestone(SpecMilestone.FULU)
+          .getConfig()
+          .toVersionFulu()
+          .orElseThrow()
+          .getMinEpochsForDataColumnSidecarsRequests();
 
   @Test
   void shouldClampResultToFuluActivationWhenCalculatedPeriodIsEarlier() {
-    final UInt64 currentEpoch = FULU_ACTIVATION_EPOCH.plus(CUSTODY_PERIOD_EPOCHS - 1);
-    final UInt64 currentSlot = currentEpoch.times(SLOTS_PER_EPOCH);
+    final UInt64 currentEpoch =
+        fuluActivationEpoch.plus(minEpochsForDataColumnSidecarsRequests - 1);
+    final UInt64 currentSlot = currentEpoch.times(slotsPerEpoch);
 
     final MinCustodyPeriodSlotCalculator calculator =
         MinCustodyPeriodSlotCalculator.createFromSpec(spec);
 
-    final UInt64 resultSlot = calculator.getMinCustodyPeriodSlot(currentSlot);
+    final Optional<UInt64> resultSlot = calculator.getMinCustodyPeriodSlot(currentSlot);
 
-    final UInt64 expectedEpoch = FULU_ACTIVATION_EPOCH; // 100
-    assertThat(resultSlot).isEqualTo(expectedEpoch.times(SLOTS_PER_EPOCH));
+    final UInt64 expectedEpoch = fuluActivationEpoch; // 100
+    assertThat(resultSlot).contains(expectedEpoch.times(slotsPerEpoch));
   }
 
   @Test
   void shouldReturnCalculatedPeriodWhenWellAfterFuluActivation() {
-    final UInt64 currentEpoch = FULU_ACTIVATION_EPOCH.plus(CUSTODY_PERIOD_EPOCHS + 10);
-    final UInt64 currentSlot = currentEpoch.times(SLOTS_PER_EPOCH);
+    final UInt64 currentEpoch =
+        fuluActivationEpoch.plus(minEpochsForDataColumnSidecarsRequests + 10);
+    final UInt64 currentSlot = currentEpoch.times(slotsPerEpoch);
 
     final MinCustodyPeriodSlotCalculator calculator =
         MinCustodyPeriodSlotCalculator.createFromSpec(spec);
 
-    final UInt64 resultSlot = calculator.getMinCustodyPeriodSlot(currentSlot);
+    final Optional<UInt64> resultSlot = calculator.getMinCustodyPeriodSlot(currentSlot);
 
     final UInt64 expectedEpoch = UInt64.valueOf(110);
-    assertThat(resultSlot).isEqualTo(expectedEpoch.times(SLOTS_PER_EPOCH));
+    assertThat(resultSlot).contains(expectedEpoch.times(slotsPerEpoch));
   }
 
   @Test
   void shouldHandleExactBoundaryAtFuluActivation() {
-    final UInt64 currentEpoch = FULU_ACTIVATION_EPOCH;
-    final UInt64 currentSlot = currentEpoch.times(SLOTS_PER_EPOCH);
+    final UInt64 currentEpoch = fuluActivationEpoch;
+    final UInt64 currentSlot = currentEpoch.times(slotsPerEpoch);
 
     final MinCustodyPeriodSlotCalculator calculator =
         MinCustodyPeriodSlotCalculator.createFromSpec(spec);
 
-    final UInt64 resultSlot = calculator.getMinCustodyPeriodSlot(currentSlot);
+    final Optional<UInt64> resultSlot = calculator.getMinCustodyPeriodSlot(currentSlot);
 
-    assertThat(resultSlot).isEqualTo(FULU_ACTIVATION_EPOCH.times(SLOTS_PER_EPOCH));
+    assertThat(resultSlot).contains(fuluActivationEpoch.times(slotsPerEpoch));
+  }
+
+  @Test
+  void shouldHandlePreFuluSlot() {
+    final UInt64 currentEpoch = fuluActivationEpoch.minus(10);
+    final UInt64 currentSlot = currentEpoch.times(slotsPerEpoch);
+
+    final MinCustodyPeriodSlotCalculator calculator =
+        MinCustodyPeriodSlotCalculator.createFromSpec(spec);
+
+    final Optional<UInt64> resultSlot = calculator.getMinCustodyPeriodSlot(currentSlot);
+
+    assertThat(resultSlot).isEmpty();
   }
 }

--- a/ethereum/statetransition/src/testFixtures/java/tech/pegasys/teku/statetransition/datacolumns/DasCustodyStand.java
+++ b/ethereum/statetransition/src/testFixtures/java/tech/pegasys/teku/statetransition/datacolumns/DasCustodyStand.java
@@ -156,7 +156,7 @@ public class DasCustodyStand {
     return custodyGroupCountManager.getCustodyColumnIndices();
   }
 
-  public UInt64 getMinCustodySlot() {
+  public Optional<UInt64> getMinCustodySlot() {
     return minCustodyPeriodSlotCalculator.getMinCustodyPeriodSlot(currentSlot);
   }
 

